impeller/entity/contents/foreground_blend_contents.cc

```diff
@@ -0,0 +1,169 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "foreground_blend_contents.h"
+#include <iostream>
+
+#include "flutter/impeller/entity/contents/content_context.h"
+#include "flutter/impeller/renderer/render_pass.h"
+#include "flutter/impeller/renderer/sampler_library.h"
+
+namespace impeller {
+
+AdvancedForegroundBlendContents::AdvancedForegroundBlendContents() {}
+
+AdvancedForegroundBlendContents::~AdvancedForegroundBlendContents() {}
+
+void AdvancedForegroundBlendContents::SetBlendMode(BlendMode blend_mode) {
+  FML_DCHECK(blend_mode > Entity::kLastPipelineBlendMode);
+  blend_mode_ = blend_mode;
+}
+
+void AdvancedForegroundBlendContents::SetSrcInput(
+    std::shared_ptr<FilterInput> input) {
+  input_ = std::move(input);
+}
+
+void AdvancedForegroundBlendContents::SetForegroundColor(Color color) {
+  foreground_color_ = color;
+}
+
+void AdvancedForegroundBlendContents::SetCoverage(Rect rect) {
+  rect_ = rect;
+}
+
+std::optional<Rect> AdvancedForegroundBlendContents::GetCoverage(
+    const Entity& entity) const {
+  return rect_.TransformBounds(entity.GetTransformation());
+}
+
+bool AdvancedForegroundBlendContents::Render(const ContentContext& renderer,
+                                             const Entity& entity,
+                                             RenderPass& pass) const {
+  using VS = BlendScreenPipeline::VertexShader;
+  using FS = BlendScreenPipeline::FragmentShader;
+
+  auto& host_buffer = pass.GetTransientsBuffer();
+
+  auto dst_snapshot = input_->GetSnapshot(renderer, entity);
+  if (!dst_snapshot.has_value()) {
+    return false;
+  }
+  auto maybe_dst_uvs = dst_snapshot->GetCoverageUVs(rect_);
+  if (!maybe_dst_uvs.has_value()) {
+    return false;
+  }
+  auto dst_uvs = maybe_dst_uvs.value();
+
+  auto size = rect_.size;
+  VertexBufferBuilder<VS::PerVertexData> vtx_builder;
```

Comment:
From reading render_filter, it appears that the coverage rect we are given already has the CTM applied. So we have two choices here: either directly use the values for the coordinates (and don't apply it to the MVP - or invert the transform and likewise.

Of course, if you don't invert the transform and fix the coordinates, then the UVs come out wrong and that's why I've "Fixed" them below.

I'm not sure exactly why this worked before

Reply:
So there is some difference in a snapshot that is created by a contents, and a snapshot that is forwarded through from an entity pass. I'm tempted just to say, for this particular case, we should just set the UVs using the coverage rect, since that coverage rect already includes the CTM it is correct for both entity subpasses and for contents. But I'm still not really sure how I should be using the snapshot GetCoverageUVs directly which is concerning.

Reply:
I guess more of a question for @bdero, but what is the correct way to use snapshot.GetCoverageUVs ?